benchmarks/smallmatvec/manual_versions/openmp/kdriver.F90

```diff
@@ -0,0 +1,669 @@
+                        cmap(1,c1_index) = (i2 * nsize) + i1 + 1
+                        c1_index = c1_index + 1
+                    else
+                        cmap(2,c2_index) = (i2 * nsize) + i1 + 1
+                        c2_index = c2_index + 1
+                    endif
+                else if (mod(i1,2).eq.0) then
+                    if (mod(i2,2).eq.0) then
+                        cmap(3,c3_index) = (i2 * nsize) + i1 + 1
+                        c3_index = c3_index + 1
+                    else
+                        cmap(4,c4_index) = (i2 * nsize) + i1 + 1
+                        c4_index = c4_index + 1
+                    endif
+                endif
+            enddo
+        enddo
+
+      else if (traverse.eq."colouring-rows") then
+          ! In this version ncp_colour contains the number of rows per color and
+          ! cmap the first cell of each coloured row
+
+        ncp_colour(1) = (nsize / 2) + mod(nsize,2)
+        ncp_colour(2) = (nsize / 2)
+        ncp_colour(3) = 0
+        ncp_colour(4) = 0
+        allocate(cmap(ncolour,maxval(ncp_colour))) !color map (colour, cell number)
+
+        c1_index = 1
+        c2_index = 1
+            do i1 = 0, nsize - 1
+                if (mod(i1,2).eq.0) then
+                        cmap(1,c1_index) = (i1 * nsize) + 1
+                        c1_index = c1_index + 1
+                else
+                        cmap(2,c2_index) = (i1 * nsize) + 1
+                        c2_index = c2_index + 1
+                endif
+            enddo
+
+      else
+          write(*,*) "Error: No colouring strategy selected"
+      endif
+      
+      ! Random data
+      call random_number(theta_adv_term_data)
+      call random_number(x_data)
+      call random_number(ptheta_2_local_stencil)
+      theta_adv_term_data = theta_adv_term_data * 100000
+      x_data = x_data * 100000
+      ptheta_2_local_stencil = ptheta_2_local_stencil * 100000
+      if (.false.) then
+              do i1 = 1, undf_any_space_1_theta_adv_term
+                theta_adv_term_data(i1) = 1
+              enddo
+              do i1=1, undf_any_space_2_x
+                x_data(i1) = 1
+              enddo
+              ptheta_2_local_stencil =1
+      endif
+      ans_data =1
+  endif
+
+  call system_mem_usage(memend)
+
+  end = omp_get_wtime()
+  write(*,*) "Allocation and initialization time =" , end - start, " s"
+  write(*,*) "Datastructure RSS memory allocated =" , memend - memstart, " KB"
+
+  ! Starting computation phase
+  write(*,*) "Starting mv with:"
+  write(*,*) "Colors=", ncolour, " -> ", ncp_colour
+  write(*,*) "Layers=", nlayers
+  write(*,*) "Horitzontal Cells=", ncell_3d/nlayers, ncell
+  write(*,*) "3D Cells=", ncell_3d
+  write(*,*) "ndf1=", ndf_any_space_1_theta_adv_term, ", undf1=", undf_any_space_1_theta_adv_term
+  write(*,*) "ndf2=", ndf_any_space_2_x, ", undf2=", undf_any_space_2_x
+
+
+  if (dotoutput.eq."write")then
+      open(1234, file="2dgrid.dot", action="write")
```

Comment:
Please add a comment describing what the created file will contain.

Reply:
I decided to delete this block, I used it to debug the generation of cell colours, but now this functionality already works and this code just works under certain conditions, so I don't think is needed anymore.

Reply:
Great.